sdk/include/opentelemetry/sdk/metrics/state/metric_storage.h

```diff
@@ -13,17 +14,20 @@ namespace metrics
 {
 
 /* Represent the storage from which to collect the metrics */
+class CollectorHandle;
 
 class MetricStorage
 {
 public:
   /* collect the metrics from this storage */
-  virtual bool Collect(
-      MetricCollector *collector,
-      nostd::span<MetricCollector *> collectors,
-      opentelemetry::sdk::instrumentationlibrary::InstrumentationLibrary *instrumentation_library,
-      opentelemetry::sdk::resource::Resource *resource,
-      nostd::function_ref<bool(MetricData)> callback) noexcept = 0;
+  virtual bool Collect(CollectorHandle *collector,
+                       nostd::span<std::shared_ptr<CollectorHandle>> collectors,
```

Comment:
Why do we need `collector` and `collectors`?

Reply:
The `collector` is the identity of MetricReader on which Collection is invoked. `collectors` are the list of all the registers MetricReaders. This is a placeholder for now and planned to use the clear internal stash once it is reported for all the MetricReaders.

Reply:
Also `collectors` would be required to maintain the unreported cumulative metrics for all the collectors separately in storage. This is mainly required because of the one of the requirements in the specs:

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#metricreader

`
The SDK MUST support multiple MetricReader instances to be registered on the same MeterProvider, and the [MetricReader.Collect](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#collect) invocation on one MetricReader instance SHOULD NOT introduce side-effects to other MetricReader instances.
`

As per the current design (inspired by JAVA implementation), the in-memory storage would maintain
 - `Delta metrics` for all the measurements received since the last collection 
 - `Unreported metrics` for all the MetricReaders.

So when a new measurement is received
 - It is aggregated and copied to the delta metrics storage.

When a new collection request is received from one of the MetricReader ( say metric_reader_1)
- The metrics are copied from the delta metric storage to the unreported metric storage of all the MetricReaders.
- Delta metric storage is cleared.
- Metrics for metric_reader_1 are fetched from Unreported metric storage and sent back.

This `collectors` would be required for the step 1 in collection above.